### PR TITLE
[MIG] contract_payment_auto: Migration from 10.0 to 12.0

### PR DESCRIPTION
**Pull request by norlinhenrik**
I need contract_payment_auto for 12.0. [How can I use this with e.g. Stripe?](https://odoo-community.org/groups/contributors-15/contributors-151468) Is there a way in Odoo community to let customers start a subscription/contract? Then the contract would recurringly generate invoices and automatic payments with e.g. Stripe.

I have NOT written any migration scripts.

All tests were disabled in [10.0](https://github.com/OCA/contract/blob/10.0/contract/__init__.py). 
I have enabled the tests, and disabled some tests because "The ['Test Acquirer'] payment acquirers are not allowed to manual capture mode!"

### CODE REVIEW

**Comment by norlinhenrik** [comment]
I did one functional change on the field is_auto_pay: default=False

File: contract_payment_auto/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+{
+    "name": "Contract - Auto Payment",
+    "summary": "Adds automatic payments to contracts.",
+    "version": "12.0.1.0.0",
+    "category": "Contract Management",
+    "license": "AGPL-3",
+    "author": "LasLabs, "
+              "Odoo Community Association (OCA)",
+    "website": "https://laslabs.com",
```
Comment (nikul-serpentcs) [inline contract_payment_auto/__manifest__.py]: @norlinhenrik URL Should be https://github.com/OCA/contract

**Review by nikul-serpentcs — COMMENTED** [review]
minor change

File: contract_payment_auto/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Copyright 2017 LasLabs Inc.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from . import models
+from . import tests
```
Comment (brian10048) [inline contract_payment_auto/__init__.py:5]:
```suggestion
```
Not required to run tests

File: contract_payment_auto/README.rst [hunk 3]
```diff
@@ -0,0 +1,96 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (brian10048) [inline contract_payment_auto/README.rst:1]: Can you separate the README sections into their own files in a README directory [as described here](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0#tasks-to-do-in-the-migration)

**Review by brian10048 — CHANGES_REQUESTED** [review]

**Comment by chafique-delli** [comment]
@norlinhenrik , can you :
- apply the changes suggested above
- check travis CI why it's failed
Thanks.

**Comment by norlinhenrik** [comment]
[https://github.com/OCA/contract/blob/12.0/contract_variable_qty_timesheet/tests/test_contract_variable_qty_timesheet.py#L49](https://github.com/OCA/contract/blob/12.0/contract_variable_qty_timesheet/tests/test_contract_variable_qty_timesheet.py#L49)
`contract._recurring_create_invoice(date_ref)`
Why does [Travis](https://travis-ci.org/github/OCA/contract/jobs/709310827#L2040) give this ERROR?
`_recurring_create_invoice() takes 1 positional argument but 2 were given`

File: contract_payment_auto/models/contract.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            invoice = invoice_line.invoice_id
+            fail_time = invoice.auto_pay_failed
+            retry_delta = timedelta(hours=contract.auto_pay_retry_hours)
+            retry_time = fail_time + retry_delta
+
+            if retry_time < now:
+                contract._do_auto_pay(invoice)
+
+    @api.multi
+    def _recurring_create_invoice(self):
```
Comment (chafique-delli) [inline contract_payment_auto/models/contract.py]: replace '_recurring_create_invoice' by 'recurring_create_invoice' method

File: contract_payment_auto/tests/test_contract.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        })
+
+
+    def _validate_invoice(self, invoice):
+        self.assertEqual(len(invoice), 1)
+        self.assertEqual(invoice._name, 'account.invoice')
+
+    def _create_invoice(self, opened=False, sent=False):
+        self.contract.is_auto_pay = False
+        invoice = self.contract._recurring_create_invoice()
```
Comment (chafique-delli) [inline contract_payment_auto/tests/test_contract.py:79]: replace '._recurring_create_invoice()' by '.recurring_create_invoice()' method

File: contract_payment_auto/tests/test_contract.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+        """ It should clear the payment token. """
+        self.assertTrue(self.contract.payment_token_id)
+        self.contract._onchange_partner_id_payment_token()
+        self.assertFalse(self.contract.payment_token_id)
+
+    def test_create_invoice_no_autopay(self):
+        """ It should return the new invoice without calling autopay. """
+        self.contract.is_auto_pay = False
+        with mock.patch.object(self.contract, '_do_auto_pay') as method:
+            invoice = self.contract._recurring_create_invoice()
```
Comment (chafique-delli) [inline contract_payment_auto/tests/test_contract.py:126]: replace '._recurring_create_invoice()' by '.recurring_create_invoice()' method

File: contract_payment_auto/tests/test_contract.py [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+            self._validate_invoice(invoice)
+            method.assert_not_called()
+
+    def test_create_invoice_autopay(self):
+        """ It should return the new invoice after calling autopay. """
+        self.contract.is_auto_pay = True
+        with mock.patch.object(self.contract, '_do_auto_pay') as method:
+            invoice = self.contract._recurring_create_invoice()
```
Comment (chafique-delli) [inline contract_payment_auto/tests/test_contract.py:134]: replace '._recurring_create_invoice()' by '.recurring_create_invoice()' method

**Review by chafique-delli — COMMENTED** [review]
@norlinhenrik , fixes to be made in the test_contract.py file:
contract_payment_auto/tests/test_contract.py:10:1: F401 'odoo.tools.mute_logger' imported but unused
contract_payment_auto/tests/test_contract.py:13:1: F401 '..models.contract' imported but unused
contract_payment_auto/tests/test_contract.py:77:5: E303 too many blank lines (2)
contract_payment_auto/tests/test_contract.py:121:5: E303 too many blank lines (2)
contract_payment_auto/tests/test_contract.py:208:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:209:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:213:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:214:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:220:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:221:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:228:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:229:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:230:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:245:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:246:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:247:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:255:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:256:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:264:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:265:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:274:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:275:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:312:5: E265 block comment should start with '# '
contract_payment_auto/tests/test_contract.py:312:89: E501 line too long (105 > 88 characters)
contract_payment_auto/tests/test_contract.py:313:5: E265 block comment should start with '# '

File: contract_payment_auto/models/contract.py (hunk 4 above)
Comment (norlinhenrik) [reply contract_payment_auto/models/contract.py]: Instead I have added `date_ref=False`

**Review by chafique-delli — APPROVED** [review]

**Review by brian10048 — APPROVED** [review]

**Comment by fcayre** [comment]
I faced several problems while testing the module port to 12.0:
- permissions for payment tokens and transactions are missing: no one can use the "Create invoices" button as only the owner of a token and transaction is allowed to read the token / transaction
- the contract module returns several invoices in _recurring_create_invoice, which makes your implementation crash; even more serious is that the returned invoices can be issued from different contracts which is_auto_pay attribute may be different and there is no more a relation from one invoice to its contract to find out the is_auto_pay value applicable to the invoice...

**Comment by legalsylvain** [comment]
Hi @norlinhenrik. What is the state of this PR ? could you take a look on @fcayre comment ? 
kind regards.

**Comment by rousseldenis** [comment]
@norlinhenrik Could we switch to #798 and close this ?

**Comment by ghost** [comment]
@rousseldenis Yes sure!

**Pull request closed without merging**